ZOOKEEPER-2684 commitProcessor does not crash when an unseen commit somes

nerdyyatrice opened the pull request:
commitProcessor with the zookeeper-2024 improvement patch throws an exception when it sees a commit request that is not at the queue head.  It turned out that it is actually a valid case when there is session movement. After discussion with the community, I submit this pull request to mitigate this issue by passing those commits to the next processor instead.

shralex commented on src/java/main/org/apache/zookeeper/server/quorum/CommitProcessor.java:
```diff
@@ -251,27 +252,29 @@ public void run() {
-                            throw new IOException("Error: unexpected cxid for"
-                                    + "client session");
+                        // we can get commit requests for requests before we even get this session
+                        //(see ZOOKEEPER-2684). We will just pass those commits to the next processor.
```
please remove the new comments from here and add them inside the if, where this case is handled.

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/CommitProcessorConcurrencyTest.java:
```diff
@@ -106,6 +106,7 @@ public void shutdown() {
 
                 @Override
                 public void notifyStopping(String threadName, int errorCode) {
+                    Assert.fail("Commit processor crashed " + errorCode);
```
Could you please adopt Kfir's test case ? its in the patch he aploaded in the JIRA.

nerdyyatrice replied:
I wonder if you have any insight on what does his test do and mine don't? I thought it was pretty much the same

shralex replied:
oh, I didn't see your test, sorry. I think there's a problem with the way you created the PR. did you run "git commit" twice ? github shows two commits here - see the commits tab. I'm also new to git, but the way I usually do it is first time the PR is created I run
> git add -u
> git commit -m 'JIRA number and description'
> git push

then if I want to change something I do:
> git add -u
> git amend
> git push -f

Now I think you should remove one of the commits from this PR. Please look online how to do that...

hanm replied:
I think the two commits are different - they changed different part of code. It is indeed confusing because they share same commit message, which is better to be fixed via git amend, but in our case it is also fine to leave them as they are because when merge the pull request the committer will squash the commits into a single commit anyway.

nerdyyatrice replied:
My first commit failed as I used lamba so I changed it back to Java 7 style. I forgot to change the commit message when I submit the second one. I will see how to amend that.

kfirlevari commented on src/java/test/org/apache/zookeeper/server/quorum/CommitProcessorConcurrencyTest.java:
```diff
@@ -374,4 +372,52 @@ public void noStarvationOfReadRequestsTest() throws Exception {
+        processor.queuedRequests.add(firstCommittedReq);
+        localRequests.add(firstCommittedReq);
+        // queue read linkedRequest to queuedRequests
+        for (; readReqId <= iteration+firstCXid; ++readReqId) {
```
I think we can just add 1 read request prior to the update request (i.e., keep the test as simple as possible and more readable). Also it'll be good to state the cxid of each new request in the actual code, so that it'll be clear what is added and when (e.g., instead of using "firstCXid", just write 0x100, 0x99, 0x98).

kfirlevari commented on src/java/test/org/apache/zookeeper/server/quorum/CommitProcessorConcurrencyTest.java:
```diff
@@ -374,4 +372,52 @@ public void noStarvationOfReadRequestsTest() throws Exception {
+        processor.initThreads(iteration* 2);
+        processor.run();
+        Assert.assertTrue(processedRequests.isEmpty());
+        // now we get a commit originated from another host that handles this same session
```
The story of the test so far wasn't clear to me until this point. Maybe add some comment stating that:
1. We add an update followed by a read operation of the same session
2. We verify that it the processor is waiting for the commit
3. We add a commit that belongs to the same session but with smaller cxid, i.e., commit of an update from previous connection of this session.   
4. We verify that the commit processor processed the old commit prior to the newer messages
5. We verify that the commit processor handle all messages.

kfirlevari commented on src/java/test/org/apache/zookeeper/server/quorum/CommitProcessorConcurrencyTest.java:
```diff
@@ -374,4 +372,52 @@ public void noStarvationOfReadRequestsTest() throws Exception {
+        processor.committedRequests.add(preSessionCommittedReq);
+        processor.committedRequests.add(firstCommittedReq);
+        processor.run();
+        Assert.assertTrue(processedRequests.peek() == preSessionCommittedReq);
```
It might be clearer if we change it a bit - only add the preSessionCommittedReq, and after you that it was handled. 
Then we add the following commit (i.e., firstCommittedReq), and verify that the rest of the messages were processed (+  verify that they weren't processed before firstCommittedReq was added).

nerdyyatrice commented:
Thank you for all the comments. However, I noticed something alarmingly bad while testing my patch and I am not if it's caused by the code change or something else. So please allow me some time to make sure that the patch actually works before addressing your comments one by one.

nerdyyatrice commented:
How can I rekick this build? The test failed at random places

hanm commented:
ZOOKEEPER-2684 kick jenkins bot

hanm commented:
@nerdyyatrice I don't see Jenkins pre-commit build is kicked off for your patch. I think a combination of these should kick the build bot:
* Put string "ZOOKEEPER-2684" in pull request title. The bot running a script that looks for JIRA number.
* And, after update title, close and reopen this pull request. This should again trigger a pre-commit build.
Do you mind do these so we can sign off this pull request from jenkins?

hanm commented:
noCrashOnCommittedRequestsOfUnSeenRequestTest is failing here, please investigate.
https://builds.apache.org/job/ZooKeeper_UT_StressTest/4152/testReport/junit/org.apache.zookeeper.server.quorum/CommitProcessorConcurrencyTest/noCrashOnCommittedRequestsOfUnSeenRequestTest/

(This is a Jenkins job I set up to stress test unit test for specific pull request branch).

nerdyyatrice commented:
Thanks Han, will do.  so do I still need to re-kick the Jenkins job?

> On May 19, 2017, at 9:44 AM, Michael Han <notifications@github.com> wrote:
> 
> noCrashOnCommittedRequestsOfUnSeenRequestTest is failing here, please investigate.
> https://builds.apache.org/job/ZooKeeper_UT_StressTest/4152/testReport/junit/org.apache.zookeeper.server.quorum/CommitProcessorConcurrencyTest/noCrashOnCommittedRequestsOfUnSeenRequestTest/ <https://builds.apache.org/job/ZooKeeper_UT_StressTest/4152/testReport/junit/org.apache.zookeeper.server.quorum/CommitProcessorConcurrencyTest/noCrashOnCommittedRequestsOfUnSeenRequestTest/>
> (This is a Jenkins job I set up to stress test unit test for specific pull request branch).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/apache/zookeeper/pull/167#issuecomment-302753487>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ALGPjBzwO5Ghp6KIMM2JNBNxO7Kc7ovSks5r7cb2gaJpZM4L85T6>.
>

hanm commented:
@nerdyyatrice I see you've closed / reopened this PR, and the build bot is kicked off as expected.  There is the same test failure in pre-commit build: https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/739/testReport/junit/org.apache.zookeeper.server.quorum/CommitProcessorConcurrencyTest/noCrashOnCommittedRequestsOfUnSeenRequestTest/. We should get this fixed before merging this PR.

I'll try reviewing this patch in detail to understand what goes wrong here.

nerdyyatrice commented:
I can’t repro it on my laptop so I tried to see if it can pass on the machine. I will look into code in more detail. 

> On May 23, 2017, at 9:15 PM, Michael Han <notifications@github.com> wrote:
> 
> @nerdyyatrice <https://github.com/nerdyyatrice> I see you've closed / reopened this PR, and the build bot is kicked off as expected. There is the same test failure in pre-commit build: https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/739/testReport/junit/org.apache.zookeeper.server.quorum/CommitProcessorConcurrencyTest/noCrashOnCommittedRequestsOfUnSeenRequestTest/ <https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/739/testReport/junit/org.apache.zookeeper.server.quorum/CommitProcessorConcurrencyTest/noCrashOnCommittedRequestsOfUnSeenRequestTest/>. We should get this fixed before merging this PR.
> 
> I'll try reviewing this patch in detail to understand what goes wrong here.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/apache/zookeeper/pull/167#issuecomment-303613061>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ALGPjMle9unhxIR4OqgDteDFY_uCDzcLks5r867ygaJpZM4L85T6>.
>

fpj commented on src/java/main/org/apache/zookeeper/server/quorum/CommitProcessor.java:
```diff
@@ -254,24 +254,23 @@ public void run() {
+                            LOG.warn("Got request " + request + 
+                                " but we are expecting request " + topPending);
+                            sessionQueue.addFirst(topPending);
+                        } else {                            
```
Is it the case that for a given session, once we execute the else block once, executing the if block would be incorrect? If so, would it make sense to have a flag per session indicating that the else block has not been executed for the session? It might not even be a flag per session, but perhaps a set of session ids instead that we remove from once we execute the else block.

anmolnar commented:
Fixed by 1392a8b3d8ac3d042a51dac18e33e144674b1b5b
